llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp

```diff
@@ -8178,6 +8178,81 @@ static Instruction *foldFCmpFSubIntoFCmp(FCmpInst &I, Instruction *LHSI,
   return nullptr;
 }
 
+static Instruction *foldFCmpWithFloorAndCeil(FCmpInst &I,
+                                             InstCombinerImpl &CI) {
+  Value *LHS = I.getOperand(0), *RHS = I.getOperand(1);
+  Type *OpType = LHS->getType();
+  CmpInst::Predicate Pred = I.getPredicate();
+
+  bool floor_x = match(LHS, m_Intrinsic<Intrinsic::floor>(m_Specific(RHS)));
+  bool x_floor = match(RHS, m_Intrinsic<Intrinsic::floor>(m_Specific(LHS)));
+  bool ceil_x = match(LHS, m_Intrinsic<Intrinsic::ceil>(m_Specific(RHS)));
+  bool x_ceil = match(RHS, m_Intrinsic<Intrinsic::ceil>(m_Specific(LHS)));
+
+  if (x_floor || x_ceil) {
+    std::swap(LHS, RHS);
+    Pred = I.getSwappedPredicate();
+    (x_floor ? floor_x : ceil_x) = true;
```

Comment:
Instead of pre-matching all the cases, you could match the preferred direction and then only try the other match if they both failed. You can also then directly assign floor_x / ceil_x from the match function when commuting.

Also variable names should be camel case